### PR TITLE
v3 - per location compensations

### PR DESCRIPTION
**Pull request by mathiazom**
Furter explanation is found here: #567 

Each `Compensations` can have one or more `location` references.

---

## Visual Overview (Image/Video)

(shows ", and more" when the number of offices reaches a limit - currently set to 3)

image

---

## Checklist

Please ensure that you’ve completed the following checkpoints before submitting your pull request:

- [x] **Documentation**: Relevant documentation has been added or updated (if applicable).
- [x] **Testing**: Have you tested your changes thoroughly?

### CODE REVIEW

**Comment by anemne** [comment]
Syntes egentlig det ser bra ut jeg! Men hvordan blir det dersom en by har flere referanser til seg?

**Comment by mathiazom** [comment]
> Syntes egentlig det ser bra ut jeg! Men hvordan blir det dersom en by har flere referanser til seg?

Godt poeng! Slik det er satt opp nå sjekker den bare at listen med kontorer ikke inneholder duplikater. For å sjekke over alle Compensations må vi legge til mer validering. Alternativt kan vi anta at man ikke kobler et kontor til flere Compensations, og bare bruke den første dersom det skjer.

File: studio/schemas/objects/office.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import { defineField } from "sanity";
+import { officeId } from "../documents/office";
+
+const office = defineField({
+  name: "office",
+  title: "Office",
```
Comment (christinaroise) [inline studio/schemas/objects/office.ts:6]: let's add a more intuitive title and description for our creators

File: studio/schemas/objects/offices.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import { defineField } from "sanity";
+import office from "./office";
+
+const offices = defineField({
+  name: "offices",
+  title: "Offices",
```
[the remark this answers is not in the document]
Comment (mathiazom) [reply studio/schemas/objects/offices.ts:6]: A little difficult to give a helpful description when the field is made reusable and included in different places. Any suggestions? Description could be "Which office is this relevant for?". We could also override the `title` and `description` props where the field is used:

```js
...
    {
      ...offices,
      description: "Which offices are these compensations relevant for?",
    },
...
```

File: studio/schemas/documents/compensations.ts [hunk 3]
```diff
@@ -29,6 +34,31 @@ const compensations = defineType({
       of: [{ type: benefitId }],
     }),
   ],
+  preview: {
+    select: {
+      subtitle: "basicTitle",
+      // https://www.sanity.io/docs/previews-list-views#62febb15a63a
```
[the remark this answers is not in the document]
Comment (mathiazom) [reply studio/schemas/documents/compensations.ts:40]: This was to help understand the fairly exotic notation for using array values in previews. Would this comment be better?

```js
/*
 Array object values are accessed with dot notation and array index: "offices.0.basicTitle"
 This limits the preview to only select a subset of the array
 https://www.sanity.io/docs/previews-list-views#62febb15a63a
*/
```

[the remark this answers is not in the document]
Comment (christinaroise) [reply studio/schemas/documents/compensations.ts:40]: Hm, upon closer inspection, how come you have this set in select? 

I also noticed that you've got the offices set as the title of the Compensation page as well: `Oslo, Trondheim, ...` 

I think a better UX would be to have:

title: "Lønn og goder" // basicTitle
subtitle: "Offices: Oslo and Trondheim"

title: "Lønn og goder" // basicTitle
subtitle: "Offices: Oslo, Trondheim, and more"

title: "Lønn og goder" // basicTitle
subtitle: "Office: Oslo"

File: studio/schemas/objects/offices.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (christinaroise) [reply studio/schemas/objects/offices.ts:6]: Good question. I'd come up with something very generic so it has one by default, and then override it when I want something more specific. 

So by default we could have title: "List of offices", "Add the office(s) you find relevant", but something much better.

File: studio/schemas/documents/compensations.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (mathiazom) [reply studio/schemas/documents/compensations.ts:40]: Where else would you get the array values than in `select`?

We can definitely swap the title and subtitle and adjust the list format :+1:

[the remark this answers is not in the document]
Comment (christinaroise) [reply studio/schemas/documents/compensations.ts:40]: We can definitely swap the title and subtitle and adjust the list format 👍

Let's do it

[the remark this answers is not in the document]
Comment (mathiazom) [reply studio/schemas/documents/compensations.ts]: Added changes and updated screenshot ✅

File: studio/schemas/objects/offices.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (mathiazom) [reply studio/schemas/objects/offices.ts:6]: Updated with default descriptions ✅

**Review by anemne — APPROVED** [review]

File: studio/schemas/objects/offices.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (christinaroise) [reply studio/schemas/objects/offices.ts:6]: I'm not so sure I agree with my previous comments. Sorry. I took a look and this is my suggestions:
```
// I would rename all things office to location
import { defineField } from "sanity";
import { officeId } from "../documents/office";

const locations = defineField({
  name: "locations",
  type: "array",
  title: "Relevant Locations",
  description:
    "You can tailor this content to specific office locations by selecting them here. If the content applies to all locations, just leave this field empty.",
  of: [
    {
      title: "Select a location",
      type: "reference",
      to: [{ type: locationId }],
      options: {
        disableNew: true,
      },
    },
  ],
});

export default locations;
```

File: studio/schemas/objects/office.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
+import { defineField } from "sanity";
```
Comment (christinaroise) [inline studio/schemas/objects/office.ts:1]: I'd omit this for now and just reference directly inside locations/offices

File: studio/schemas/documents/compensations.ts [hunk 5]
```diff
@@ -1,16 +1,31 @@
...
 export const compensationsId = "compensations";
 
 const compensations = defineType({
   name: compensationsId,
   type: "document",
   title: "Compensations",
   fields: [
+    {
+      ...offices,
+      description: "Which offices are these compensations relevant for?",
```
Comment (christinaroise) [inline studio/schemas/documents/compensations.ts:27]: I like what you had before. I think we can adjust if we see fit later. I added suggestions on title and descriptions down below 👇

File: studio/schemas/documents/compensations.ts [hunk 6]
```diff
@@ -1,16 +1,31 @@
...
 
+/*
+ maximum number of offices without preview string shortening
+ (should be at least 2)
+ given array: ["Trondheim", "Oslo", "Bergen", "Molde"]
+ 2 => Trondheim, and more
+ 3 => Trondheim, Oslo, and more
+ 4 => Trondheim, Oslo, Bergen, and Molde
+*/
+const OFFICES_PREVIEW_CUTOFF = 3;
```
Comment (christinaroise) [inline studio/schemas/documents/compensations.ts:16]: I'd keep it more simple:
```suggestion
export const compensationsId = "compensations";  // remember to keep sorting imports -> exports -> consts 

// Defines the maximum number of office locations to display in the preview without truncating
const OFFICES_PREVIEW_CUTOFF = 3;
```

File: studio/schemas/documents/compensations.ts [hunk 7]
```diff
@@ -29,6 +44,48 @@ const compensations = defineType({
...
     }),
   ],
+  preview: {
+    select: {
+      title: "basicTitle",
+      /*
+       Array object values are accessed with dot notation and array index: "offices.0.basicTitle"
+       This limits the preview to only select a subset of the array
+       https://www.sanity.io/docs/previews-list-views#62febb15a63a
+      */
```
Comment (christinaroise) [inline studio/schemas/documents/compensations.ts:54]: I'd keep it more simple:

```suggestion
      // Dynamically selects office names for preview based on the OFFICES_PREVIEW_CUTOFF
```

File: studio/schemas/documents/compensations.ts [hunk 8]
```diff
@@ -29,6 +44,48 @@ const compensations = defineType({
...
+      return {
+        title,
+        subtitle: previewStringFromOfficesMap(
+          officesMap,
+          OFFICES_PREVIEW_CUTOFF,
+        ),
+      };
+    },
+  },
 });
```
Comment (christinaroise) [inline studio/schemas/documents/compensations.ts:71]: I'd add this:

```
/**
 * Generates a preview string based on the selected office locations.
 *
 * @param {Object} officesMap - A map of office titles selected for preview, e.g., { office0: "Trondheim", office1: "Oslo", ... }.
 * @param {number} cutoff - The maximum number of offices to display before shortening the preview string.
 * @returns {string|undefined} - A formatted string summarizing the selected offices or `undefined` if no offices are selected.
 */
 ```

File: studio/schemas/documents/compensations.ts [hunk 9]
```diff
@@ -1,16 +1,28 @@
...
 import { benefitId } from "./benefit";
 
+/*
+ maximum number of locations without preview string shortening
+ (should be at least 2)
+ given array: ["Trondheim", "Oslo", "Bergen", "Molde"]
+ 2 => Trondheim, and more
+ 3 => Trondheim, Oslo, and more
+ 4 => Trondheim, Oslo, Bergen, and Molde
+*/
```
Comment (christinaroise) [inline studio/schemas/documents/compensations.ts:15]: I don't see my comment anymore, but:

```suggestion
export const compensationsId = "compensations";

// Defines the maximum number of office locations to display in the preview without truncating
const OFFICES_PREVIEW_CUTOFF = 3;
```

File: studio/schemas/documents/compensations.ts (hunk 7 above)
Comment (mathiazom) [reply studio/schemas/documents/compensations.ts:51]: I think the original one is much more useful. The simpler one provides very little insight. It's hard to predict what a future developer will need, but to me the syntax was unexpected and called for some extra context.

File: studio/schemas/documents/compensations.ts (hunk 9 above)
Comment (mathiazom) [reply studio/schemas/documents/compensations.ts]: If we want consistent sorting it should be integrated with Prettier. 

How do you handle `export default` if `export` should come after `import`s?

File: studio/schemas/documents/compensations.ts (hunk 7 above)
Comment (christinaroise) [reply studio/schemas/documents/compensations.ts:44]:
```suggestion
/*
   Access array object values using dot notation with the array index, e.g., "locations.0.basicTitle".
   This approach allows selecting a subset of the array for preview purposes.
   For more details, see: https://www.sanity.io/docs/previews-list-views#62febb15a63a
*/
```

**Pull request merged**